internal/controllers/machine/machine_controller.go

```diff
@@ -791,38 +803,89 @@ func shouldRequeueDrain(now time.Time, lastDrain time.Time) (time.Duration, bool
 // this could cause issue for some storage provisioner, for example, vsphere-volume this is problematic
 // because if the node is deleted before detach success, then the underline VMDK will be deleted together with the Machine
 // so after node draining we need to check if all volumes are detached before deleting the node.
-func (r *Reconciler) shouldWaitForNodeVolumes(ctx context.Context, cluster *clusterv1.Cluster, nodeName string) (bool, error) {
+func (r *Reconciler) shouldWaitForNodeVolumes(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+	nodeName := machine.Status.NodeRef.Name
 	log := ctrl.LoggerFrom(ctx, "Node", klog.KRef("", nodeName))
 	ctx = ctrl.LoggerInto(ctx, log)
 
 	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
 	if err != nil {
-		return true, err
+		return ctrl.Result{}, err
 	}
 
 	node := &corev1.Node{}
 	if err := remoteClient.Get(ctx, types.NamespacedName{Name: nodeName}, node); err != nil {
 		if apierrors.IsNotFound(err) {
 			log.Error(err, "Could not find node from noderef, it may have already been deleted")
-			return false, nil
+			return ctrl.Result{}, nil
 		}
-		return true, err
+		return ctrl.Result{}, err
 	}
 
 	if noderefutil.IsNodeUnreachable(node) {
 		// If a node is unreachable, we can't detach the volume.
 		// We need to skip the detachment as we otherwise block deletions
 		// of unreachable nodes when a volume is attached.
 		log.Info("Node is unreachable, skip waiting for volume detachment.")
-		return false, nil
+		return ctrl.Result{}, nil
 	}
 
-	if len(node.Status.VolumesAttached) != 0 {
-		log.Info("Waiting for Node volumes to be detached")
-		return true, nil
+	// Check waitForVolumeDetach cache to ensure we won't retry before waitForVolumeDetachInterval.
+	// Note: This is intentionally only done if we have Pods to evict, because otherwise
+	// even the "no-op" drain would be requeued.
+	if cacheEntry, ok := r.waitForVolumeDetachCache.Has(client.ObjectKeyFromObject(machine)); ok {
+		if requeueAfter, requeue := shouldRequeue(time.Now(), cacheEntry.LastProceed, waitForVolumeDetachInterval); requeue {
+			log.Info(fmt.Sprintf("Requeuing in %s because there already was a wiat for node volumes in the last %s", requeueAfter, waitForVolumeDetachInterval))
```

Comment:
```suggestion
			log.Info(fmt.Sprintf("Requeuing in %s because we already checked if there are still attached volumes in the last %s", requeueAfter.Truncate(time.Second/10), waitForVolumeDetachInterval))
```
(I also added truncate for drain in the ClusterCache PR, otherwise we end up with a lot of digits)